pkg/controller/control.go

```diff
@@ -656,6 +660,78 @@ func (c *Controller) checkUnmapMarkSnapChainRemoved() error {
 	return nil
 }
 
+func (c *Controller) SetMaximumSnapshotCount(count int) error {
+	c.Lock()
+	defer c.Unlock()
+
+	c.maximumSnapshotCount = count
+
+	allFailed := true
+	for _, r := range c.replicas {
+		// The related backend is nil if the mode is ERR
+		if r.Mode == types.ERR {
+			continue
+		}
+		err := c.backend.SetMaximumSnapshotCount(r.Address, count)
+		if err != nil {
+			logrus.Errorf("Failed to set flag MaximumSnapshotCount to %d in replica %s. Mark as ERR", count, r.Address)
+			c.setReplicaModeNoLock(r.Address, types.ERR)
```

Comment:
Will setting the replica to ERR a bit aggressive? This will trigger a rebuild which is a costly operation. Is it ok to just print out the log and return ERR so that engine can retry next time? 

cc @innobead  @shuo-wu @derekbit

Reply:
Agreed.

I believe Poan just used the same implementation from other operations to apply here; however, for others, they are intended to fail the replica.

Here, I would suggest doing a 2-phase commit. Check all downstream are OK to set. If yes, the second commit will be applied. The same for MaxSnapshotSize.

Reply:
For now, I think 2-phase commit is a little overengineering. We need to acquire lock for each backend in the first commit and apply the change in the second commit. Our request it to send an RPC call. If there is any error, we need to have a way to release the lock. There're too many cases need to handle. I would prefer to return an error and reconcile it next time.

Reply:
Sounds good to me.